File: tdz/src/reminder.rs
use crate::{models::*, error::*};
use std::collections::HashMap;
use chrono::{Utc, DateTime, Duration};
pub struct ReminderManager {
    pub reminders: HashMap<String, Reminder>,
    pub reminder_tags: HashMap<String, Vec<String>>,
}
impl ReminderManager {
    pub fn new() -> Self {
        Self {
            reminders: HashMap::new(),
            reminder_tags: HashMap::new(),
        }
    }
    pub async fn create_reminder(&mut self, mut reminder: Reminder) -> Result<String> {
        reminder.id = uuid::Uuid::new_v4().to_string();
        reminder.created_at = Utc::now();
        reminder.updated_at = Utc::now();
        self.reminder_tags.insert(reminder.id.clone(), reminder.tags.clone());
        self.reminders.insert(reminder.id.clone(), reminder.clone());
        Ok(reminder.id)
    }
    pub fn get_reminder(&self, reminder_id: &str) -> Option<&Reminder> {
        self.reminders.get(reminder_id)
    }
    pub fn get_all_reminders(&self) -> Vec<&Reminder> {
        self.reminders.values().collect()
    }
    pub async fn update_reminder(
        &mut self,
        reminder_id: &str,
        updates: ReminderUpdate,
    ) -> Result<()> {
        if let Some(reminder) = self.reminders.get_mut(reminder_id) {
            if let Some(content) = updates.content {
                reminder.content = content;
            }
            if let Some(remind_at) = updates.remind_at {
                reminder.remind_at = remind_at;
            }
            if let Some(priority) = updates.priority {
                reminder.priority = priority;
            }
            if let Some(status) = updates.status {
                reminder.status = status;
            }
            if let Some(tags) = updates.tags {
                reminder.tags = tags.clone();
                self.reminder_tags.insert(reminder_id.to_string(), tags);
            }
            reminder.updated_at = Utc::now();
        } else {
            return Err(TodoziError::ValidationError {
                message: format!("Reminder {} not found", reminder_id),
            });
        }
        Ok(())
    }
    pub async fn delete_reminder(&mut self, reminder_id: &str) -> Result<()> {
        if self.reminders.remove(reminder_id).is_some() {
            self.reminder_tags.remove(reminder_id);
            Ok(())
        } else {
            Err(TodoziError::ValidationError {
                message: format!("Reminder {} not found", reminder_id),
            })
        }
    }
    pub fn search_reminders(&self, query: &str) -> Vec<&Reminder> {
        let query_lower = query.to_lowercase();
        self.reminders
            .values()
            .filter(|reminder| {
                reminder.content.to_lowercase().contains(&query_lower)
                    || reminder
                        .tags
                        .iter()
                        .any(|tag| tag.to_lowercase().contains(&query_lower))
            })
            .collect()
    }
    pub fn get_reminders_by_priority(
        &self,
        priority: ReminderPriority,
    ) -> Vec<&Reminder> {
        self.reminders
            .values()
            .filter(|reminder| reminder.priority == priority)
            .collect()
    }
    pub fn get_reminders_by_status(&self, status: ReminderStatus) -> Vec<&Reminder> {
        self.reminders.values().filter(|reminder| reminder.status == status).collect()
    }
    pub fn get_reminders_by_tag(&self, tag: &str) -> Vec<&Reminder> {
        let tag_lower = tag.to_lowercase();
        self.reminders
            .values()
            .filter(|reminder| {
                reminder.tags.iter().any(|t| t.to_lowercase() == tag_lower)
            })
            .collect()
    }
    pub fn get_pending_reminders(&self) -> Vec<&Reminder> {
        self.get_reminders_by_status(ReminderStatus::Pending)
    }
    pub fn get_active_reminders(&self) -> Vec<&Reminder> {
        self.get_reminders_by_status(ReminderStatus::Active)
    }
    pub fn get_overdue_reminders(&self) -> Vec<&Reminder> {
        let now = Utc::now();
        self.reminders
            .values()
            .filter(|reminder| {
                reminder.remind_at < now
                    && (reminder.status == ReminderStatus::Pending
                        || reminder.status == ReminderStatus::Active)
            })
            .collect()
    }
    pub fn get_reminders_due_soon(&self, duration: Duration) -> Vec<&Reminder> {
        let now = Utc::now();
        let due_time = now + duration;
        self.reminders
            .values()
            .filter(|reminder| {
                reminder.remind_at <= due_time && reminder.remind_at > now
                    && (reminder.status == ReminderStatus::Pending
                        || reminder.status == ReminderStatus::Active)
            })
            .collect()
    }
    pub fn get_recent_reminders(&self, limit: usize) -> Vec<&Reminder> {
        let mut reminders: Vec<&Reminder> = self.reminders.values().collect();
        reminders.sort_by(|a, b| b.created_at.cmp(&a.created_at));
        reminders.into_iter().take(limit).collect()
    }
    pub fn get_all_tags(&self) -> Vec<String> {
        let mut all_tags = std::collections::HashSet::new();
        for tags in self.reminder_tags.values() {
            for tag in tags {
                all_tags.insert(tag.clone());
            }
        }
        all_tags.into_iter().collect()
    }
    pub fn get_tag_statistics(&self) -> HashMap<String, usize> {
        let mut stats = HashMap::new();
        for tags in self.reminder_tags.values() {
            for tag in tags {
                *stats.entry(tag.clone()).or_insert(0) += 1;
            }
        }
        stats
    }
    pub fn get_reminder_statistics(&self) -> ReminderStatistics {
        let total_reminders = self.reminders.len();
        let pending = self.get_pending_reminders().len();
        let active = self.get_active_reminders().len();
        let overdue = self.get_overdue_reminders().len();
        let unique_tags = self.get_all_tags().len();
        ReminderStatistics {
            total_reminders,
            pending_reminders: pending,
            active_reminders: active,
            overdue_reminders: overdue,
            unique_tags,
        }
    }
    pub async fn mark_reminder_completed(&mut self, reminder_id: &str) -> Result<()> {
        if let Some(reminder) = self.reminders.get_mut(reminder_id) {
            reminder.mark_completed();
            Ok(())
        } else {
            Err(TodoziError::ValidationError {
                message: format!("Reminder {} not found", reminder_id),
            })
        }
    }
    pub async fn mark_reminder_cancelled(&mut self, reminder_id: &str) -> Result<()> {
        if let Some(reminder) = self.reminders.get_mut(reminder_id) {
            reminder.mark_cancelled();
            Ok(())
        } else {
            Err(TodoziError::ValidationError {
                message: format!("Reminder {} not found", reminder_id),
            })
        }
    }
    pub async fn activate_reminder(&mut self, reminder_id: &str) -> Result<()> {
        if let Some(reminder) = self.reminders.get_mut(reminder_id) {
            reminder.activate();
            Ok(())
        } else {
            Err(TodoziError::ValidationError {
                message: format!("Reminder {} not found", reminder_id),
            })
        }
    }
}
#[derive(Debug, Clone)]
pub struct ReminderUpdate {
    pub content: Option<String>,
    pub remind_at: Option<DateTime<Utc>>,
    pub priority: Option<ReminderPriority>,
    pub status: Option<ReminderStatus>,
    pub tags: Option<Vec<String>>,
}
impl ReminderUpdate {
    pub fn new() -> Self {
        Self {
            content: None,
            remind_at: None,
            priority: None,
            status: None,
            tags: None,
        }
    }
    pub fn content(mut self, content: String) -> Self {
        self.content = Some(content);
        self
    }
    pub fn remind_at(mut self, remind_at: DateTime<Utc>) -> Self {
        self.remind_at = Some(remind_at);
        self
    }
    pub fn priority(mut self, priority: ReminderPriority) -> Self {
        self.priority = Some(priority);
        self
    }
    pub fn status(mut self, status: ReminderStatus) -> Self {
        self.status = Some(status);
        self
    }
    pub fn tags(mut self, tags: Vec<String>) -> Self {
        self.tags = Some(tags);
        self
    }
}
#[derive(Debug, Clone)]
pub struct ReminderStatistics {
    pub total_reminders: usize,
    pub pending_reminders: usize,
    pub active_reminders: usize,
    pub overdue_reminders: usize,
    pub unique_tags: usize,
}
impl ReminderStatistics {
    pub fn pending_percentage(&self) -> f64 {
        if self.total_reminders == 0 {
            0.0
        } else {
            (self.pending_reminders as f64 / self.total_reminders as f64) * 100.0
        }
    }
    pub fn active_percentage(&self) -> f64 {
        if self.total_reminders == 0 {
            0.0
        } else {
            (self.active_reminders as f64 / self.total_reminders as f64) * 100.0
        }
    }
    pub fn overdue_percentage(&self) -> f64 {
        if self.total_reminders == 0 {
            0.0
        } else {
            (self.overdue_reminders as f64 / self.total_reminders as f64) * 100.0
        }
    }
}
pub fn parse_reminder_format(reminder_text: &str) -> Result<Reminder> {
    let start_tag = "<reminder>";
    let end_tag = "</reminder>";
    let start = reminder_text
        .find(start_tag)
        .ok_or_else(|| TodoziError::ValidationError {
            message: "Missing <reminder> start tag".to_string(),
        })?;
    let end = reminder_text
        .find(end_tag)
        .ok_or_else(|| TodoziError::ValidationError {
            message: "Missing </reminder> end tag".to_string(),
        })?;
    let content = &reminder_text[start + start_tag.len()..end];
    let parts: Vec<&str> = content.split(';').map(|s| s.trim()).collect();
    if parts.len() < 3 {
        return Err(TodoziError::ValidationError {
            message: "Invalid reminder format: need at least 3 parts (content; remind_at; priority)"
                .to_string(),
        });
    }
    let remind_at = parts[1]
        .parse::<DateTime<Utc>>()
        .map_err(|_| TodoziError::ValidationError {
            message: "Invalid reminder date format".to_string(),
        })?;
    let priority = parts[2]
        .parse::<ReminderPriority>()
        .map_err(|_| TodoziError::ValidationError {
            message: "Invalid reminder priority".to_string(),
        })?;
    let status = if parts.len() > 3 && !parts[3].is_empty() {
        parts[3]
            .parse::<ReminderStatus>()
            .map_err(|_| TodoziError::ValidationError {
                message: "Invalid reminder status".to_string(),
            })?
    } else {
        ReminderStatus::Pending
    };
    let tags = if parts.len() > 4 && !parts[4].is_empty() {
        parts[4].split(',').map(|s| s.trim().to_string()).collect()
    } else {
        Vec::new()
    };
    Ok(Reminder {
        id: uuid::Uuid::new_v4().to_string(),
        content: parts[0].to_string(),
        remind_at,
        priority,
        status,
        tags,
        created_at: Utc::now(),
        updated_at: Utc::now(),
    })
}
#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn test_reminder_manager_creation() {
        let manager = ReminderManager::new();
        assert_eq!(manager.reminders.len(), 0);
        assert_eq!(manager.reminder_tags.len(), 0);
    }
    #[test]
    fn test_reminder_update_builder() {
        let update = ReminderUpdate::new()
            .content("New content".to_string())
            .priority(ReminderPriority::High);
        assert_eq!(update.content, Some("New content".to_string()));
        assert_eq!(update.priority, Some(ReminderPriority::High));
    }
    #[test]
    fn test_reminder_statistics() {
        let stats = ReminderStatistics {
            total_reminders: 10,
            pending_reminders: 5,
            active_reminders: 3,
            overdue_reminders: 2,
            unique_tags: 4,
        };
        assert_eq!(stats.pending_percentage(), 50.0);
        assert_eq!(stats.active_percentage(), 30.0);
        assert_eq!(stats.overdue_percentage(), 20.0);
        let empty_stats = ReminderStatistics {
            total_reminders: 0,
            pending_reminders: 0,
            active_reminders: 0,
            overdue_reminders: 0,
            unique_tags: 0,
        };
        assert_eq!(empty_stats.pending_percentage(), 0.0);
        assert_eq!(empty_stats.active_percentage(), 0.0);
        assert_eq!(empty_stats.overdue_percentage(), 0.0);
    }
    #[test]
    fn test_parse_reminder_format() {
        let reminder_text = "<reminder>Review project proposal; 2025-01-20T10:00:00Z; high; pending; review,project,deadline</reminder>";
        let reminder = parse_reminder_format(reminder_text).unwrap();
        assert_eq!(reminder.content, "Review project proposal");
        assert_eq!(reminder.priority, ReminderPriority::High);
        assert_eq!(reminder.status, ReminderStatus::Pending);
        assert_eq!(reminder.tags, vec!["review", "project", "deadline"]);
    }
    #[test]
    fn test_parse_reminder_format_minimal() {
        let reminder_text = "<reminder>Simple reminder; 2025-01-20T10:00:00Z; medium</reminder>";
        let reminder = parse_reminder_format(reminder_text).unwrap();
        assert_eq!(reminder.content, "Simple reminder");
        assert_eq!(reminder.priority, ReminderPriority::Medium);
        assert_eq!(reminder.status, ReminderStatus::Pending);
        assert_eq!(reminder.tags.len(), 0);
    }
}

File: tdz/py/reminder.py
"""Reminder management for Todozi."""

from typing import Dict, List, Optional
from datetime import datetime, timedelta
import uuid as uuid_lib

from .models import Reminder, ReminderPriority, ReminderStatus
from .error import ValidationError


class ReminderManager:
    """Manages reminders."""

    def __init__(self):
        self.reminders: Dict[str, Reminder] = {}
        self.reminder_tags: Dict[str, List[str]] = {}

    async def create_reminder(self, reminder: Reminder) -> str:
        """Create a new reminder."""
        reminder.id = str(uuid_lib.uuid4())
        reminder.created_at = datetime.utcnow()
        reminder.updated_at = datetime.utcnow()
        self.reminder_tags[reminder.id] = reminder.tags.copy()
        self.reminders[reminder.id] = reminder
        return reminder.id

    def get_reminder(self, reminder_id: str) -> Optional[Reminder]:
        """Get a reminder by ID."""
        return self.reminders.get(reminder_id)

    def get_all_reminders(self) -> List[Reminder]:
        """Get all reminders."""
        return list(self.reminders.values())

    async def update_reminder(self, reminder_id: str, updates: 'ReminderUpdate'):
        """Update a reminder."""
        reminder = self.reminders.get(reminder_id)
        if not reminder:
            raise ValidationError(f"Reminder {reminder_id} not found")

        if updates.content is not None:
            reminder.content = updates.content
        if updates.remind_at is not None:
            reminder.remind_at = updates.remind_at
        if updates.priority is not None:
            reminder.priority = updates.priority
        if updates.status is not None:
            reminder.status = updates.status
        if updates.tags is not None:
            reminder.tags = updates.tags.copy()
            self.reminder_tags[reminder_id] = updates.tags

        reminder.updated_at = datetime.utcnow()

    async def delete_reminder(self, reminder_id: str):
        """Delete a reminder."""
        if reminder_id not in self.reminders:
            raise ValidationError(f"Reminder {reminder_id} not found")
        del self.reminders[reminder_id]
        self.reminder_tags.pop(reminder_id, None)

    def search_reminders(self, query: str) -> List[Reminder]:
        """Search for reminders."""
        query_lower = query.lower()
        return [
            r for r in self.reminders.values()
            if query_lower in r.content.lower()
            or any(query_lower in tag.lower() for tag in r.tags)
        ]

    def get_reminders_by_priority(self, priority: ReminderPriority) -> List[Reminder]:
        """Get reminders by priority."""
        return [r for r in self.reminders.values() if r.priority == priority]

    def get_reminders_by_status(self, status: ReminderStatus) -> List[Reminder]:
        """Get reminders by status."""
        return [r for r in self.reminders.values() if r.status == status]

    def get_reminders_by_tag(self, tag: str) -> List[Reminder]:
        """Get reminders by tag."""
        tag_lower = tag.lower()
        return [
            r for r in self.reminders.values()
            if any(t.lower() == tag_lower for t in r.tags)
        ]

    def get_pending_reminders(self) -> List[Reminder]:
        """Get pending reminders."""
        return self.get_reminders_by_status(ReminderStatus.PENDING)

    def get_active_reminders(self) -> List[Reminder]:
        """Get active reminders."""
        return self.get_reminders_by_status(ReminderStatus.ACTIVE)

    def get_overdue_reminders(self) -> List[Reminder]:
        """Get overdue reminders."""
        now = datetime.utcnow()
        return [
            r for r in self.reminders.values()
            if r.remind_at < now
            and r.status in [ReminderStatus.PENDING, ReminderStatus.ACTIVE]
        ]

    def get_reminders_due_soon(self, duration: timedelta) -> List[Reminder]:
        """Get reminders due soon."""
        now = datetime.utcnow()
        due_time = now + duration
        return [
            r for r in self.reminders.values()
            if now < r.remind_at <= due_time
            and r.status in [ReminderStatus.PENDING, ReminderStatus.ACTIVE]
        ]

    def get_recent_reminders(self, limit: int) -> List[Reminder]:
        """Get recent reminders."""
        reminders = sorted(self.reminders.values(), key=lambda x: x.created_at, reverse=True)
        return reminders[:limit]

    def get_all_tags(self) -> List[str]:
        """Get all unique tags."""
        all_tags = set()
        for tags in self.reminder_tags.values():
            all_tags.update(tags)
        return list(all_tags)

    def get_tag_statistics(self) -> Dict[str, int]:
        """Get tag usage statistics."""
        stats = {}
        for tags in self.reminder_tags.values():
            for tag in tags:
                stats[tag] = stats.get(tag, 0) + 1
        return stats

    def get_reminder_statistics(self) -> 'ReminderStatistics':
        """Get reminder statistics."""
        return ReminderStatistics(
            total_reminders=len(self.reminders),
            pending_reminders=len(self.get_pending_reminders()),
            active_reminders=len(self.get_active_reminders()),
            overdue_reminders=len(self.get_overdue_reminders()),
            unique_tags=len(self.get_all_tags())
        )

    async def mark_reminder_completed(self, reminder_id: str):
        """Mark a reminder as completed."""
        reminder = self.reminders.get(reminder_id)
        if not reminder:
            raise ValidationError(f"Reminder {reminder_id} not found")
        reminder.mark_completed()

    async def mark_reminder_cancelled(self, reminder_id: str):
        """Mark a reminder as cancelled."""
        reminder = self.reminders.get(reminder_id)
        if not reminder:
            raise ValidationError(f"Reminder {reminder_id} not found")
        reminder.mark_cancelled()

    async def activate_reminder(self, reminder_id: str):
        """Activate a reminder."""
        reminder = self.reminders.get(reminder_id)
        if not reminder:
            raise ValidationError(f"Reminder {reminder_id} not found")
        reminder.activate()


class ReminderUpdate:
    """Reminder update structure."""
    def __init__(self):
        self.content: Optional[str] = None
        self.remind_at: Optional[datetime] = None
        self.priority: Optional[ReminderPriority] = None
        self.status: Optional[ReminderStatus] = None
        self.tags: Optional[List[str]] = None


class ReminderStatistics:
    """Reminder statistics."""
    def __init__(self, total_reminders: int, pending_reminders: int,
                 active_reminders: int, overdue_reminders: int, unique_tags: int):
        self.total_reminders = total_reminders
        self.pending_reminders = pending_reminders
        self.active_reminders = active_reminders
        self.overdue_reminders = overdue_reminders
        self.unique_tags = unique_tags

    def pending_percentage(self) -> float:
        """Get percentage of pending reminders."""
        return (self.pending_reminders / self.total_reminders * 100.0) if self.total_reminders > 0 else 0.0

    def active_percentage(self) -> float:
        """Get percentage of active reminders."""
        return (self.active_reminders / self.total_reminders * 100.0) if self.total_reminders > 0 else 0.0

    def overdue_percentage(self) -> float:
        """Get percentage of overdue reminders."""
        return (self.overdue_reminders / self.total_reminders * 100.0) if self.total_reminders > 0 else 0.0


def parse_reminder_format(reminder_text: str) -> Reminder:
    """Parse reminder from XML-like format."""
    start_tag = "<reminder>"
    end_tag = "</reminder>"

    start = reminder_text.find(start_tag)
    if start == -1:
        raise ValidationError("Missing <reminder> start tag")

    end = reminder_text.find(end_tag)
    if end == -1:
        raise ValidationError("Missing </reminder> end tag")

    content = reminder_text[start + len(start_tag):end]
    parts = [s.strip() for s in content.split(';')]

    if len(parts) < 3:
        raise ValidationError(
            "Invalid reminder format: need at least 3 parts (content; remind_at; priority)"
        )

    # Parse datetime
    try:
        remind_at = datetime.fromisoformat(parts[1].replace('Z', '+00:00'))
    except:
        raise ValidationError("Invalid reminder date format")

    # Parse priority
    priority = ReminderPriority(parts[2].lower())

    # Parse status
    status = ReminderStatus.PENDING
    if len(parts) > 3 and parts[3]:
        status = ReminderStatus(parts[3].lower())

    # Parse tags
    tags = []
    if len(parts) > 4 and parts[4]:
        tags = [s.strip() for s in parts[4].split(',')]

    return Reminder(
        id=str(uuid_lib.uuid4()),
        content=parts[0],
        remind_at=remind_at,
        priority=priority,
        status=status,
        tags=tags,
        created_at=datetime.utcnow(),
        updated_at=datetime.utcnow()
    )
